projects/evm-compatible-bridge/evm_contract_design.md

```diff
@@ -0,0 +1,141 @@
+// To use Hathor native token (HTR) we can use the local id 1 which is reserved for this.
+```
+
+To check if the token is native to Hathor or the EVM we can check the contract address of the token, since all Hathor native tokens will create an EVM token on the bridge address.
+
+#### Federation methods
+
+Some methods can only be called by the federation contract (or admin).
+These methods will be used to manage tokens, include support for new tokens, etc.
+
+The federation will also provide a method so the bridge contract can notify when a user requests support for a token or a crossing of a token to Hathor.
+
+### Federation contract
+
+A good example implementation can be found [here](https://github.com/onepercentio/tokenbridge/blob/master/bridge/contracts/Federation.sol).
+This contract has a very good example of a federation with all needed functionalities, it also includes an owner account for the federation that can change the members and the bridge being managed.
+
+The main method of the federation is `voteProposal` which will be used to vote on a request.
+Requests can be used to unlock tokens to a user, to support a new token, burn or mint tokens.
+Each request will be saved with an id so it can be voted on and later checked if it was accepted or not, the id will be a hash of the request data.
+
+The federation contract will gather votes for requests and when a request has reached the amount of votes required it will execute the request calling the necessary methods on the bridge contract.
+The request will be marked as executed and no more votes will be accepted.
+
+The federation contract should also be "pausable", where we suspend the `voteProposal` method but keep admin methods, e.g. add and remove members, change the managed bridge, etc.
+
+# Rationale and alternatives
+
+_Why use events for communication?_
+
+An event (also called a log entry) is a cheaper way to store data on-chain.
+The price to allocate a 4 byte word on storage is 20.000 gas.
+In contrast an example of log gas for 2 topics and 200 bytes of data would cost 3.325 gas, 50 times more data for almost 10 times less gas.
+The method to calculate log gas can be found [here](https://github.com/ethereum/go-ethereum/blob/8a24b563312a0ab0a808770e464c5598ab7e35ea/core/vm/gas_table.go#L220).
+
+Cost is not the only benefit, responding to events is easier to implement than having getter methods for all types of required data.
+
+_Why use ERC-1155?_
+
+An alternative would be to use a SideTokenFactory contract (example [here](https://github.com/onepercentio/tokenbridge/blob/master/bridge/contracts/SideTokenFactory.sol)) but I believe we should strive to use the most common standards.
+
+# Prior art
+
+Although [this bridge](https://github.com/onepercentio/tokenbridge) implements a bridge between 2 EVM compatible chains, the concepts implemented there are a good reference for security and best practices.
```

Comment:
It is very close to what we need indeed, I made some adjustments to be more generic and allow more tokens but depending on the business requirements we can use this as a base for our contract.

This project uses 4 contracts as a base instead of 2:
- Federation
- Bridge
  - Instead of using ERC-1155 it deploys a new ERC-777 contract (using the SideToken contract) when creating a token for the first time, this may increate the cost of accepting new tokens.
  - Only supports ERC-777 (and ERC-20 due to backwards compatibility)
- MultiSig
  - This is used as the admin so any admin operation requires multiple consenting parties.
  - Could be used on the EVM side without changes to the original.
- AllowTokens
  - A contract to configure a list of allowed tokens
  - It also serves as a way to limit interactions with the bridge (by configuring a daily limit of tokens to be sent, etc.)
 
There is also another issue with how a "crossing" is done, as soon as the event is emitted it is considered as a success, we can adopt this reasoning but I think that once the transaction is sent on Hathor we should add the confirmation on the EVM chain (with the tx_id) so we can later inspect the bridge history and confirm that is matches Hathor history.
Either way I think the project can be used as a base, that is why I included the link in the design.